helix-view/src/document.rs

```diff
@@ -265,6 +276,174 @@ impl Document {
+
+                    _ => {
+                        prev_line_indent = (false, 0usize);
+                        continue 'outer;
+                    }
+                };
+
+                // Count the total leading tab/space characters.
+                let mut count = 1;
+                let mut count_is_done = false;
+                for c in c_iter {
+                    match c {
+                        '\t' if is_tabs && !count_is_done => count += 1,
+                        ' ' if !is_tabs && !count_is_done => count += 1,
+
+                        // We stop counting if we hit whitespace that doesn't
+                        // qualify as indent or doesn't match the leading
+                        // whitespace, but we don't exit the loop yet because
+                        // we still want to determine if the line is blank.
+                        c if char_is_whitespace(c) => count_is_done = true,
+
+                        // Ignore blank lines.
+                        c if char_is_linebreak(c) => continue 'outer,
+
+                        _ => break,
+                    }
+
+                    // Bound the worst-case execution time for weird text files.
+                    if count > 256 {
+                        continue 'outer;
+                    }
+                }
+
+                // Update stats.
+                if (prev_line_indent.0 == is_tabs || prev_line_indent.1 == 0)
+                    && prev_line_indent.1 < count
+                {
+                    if is_tabs {
+                        histogram[0] += 1;
+                    } else {
+                        let amount = count - prev_line_indent.1;
+                        if amount <= 8 {
+                            histogram[amount] += 1;
+                        }
+                    }
+                }
+
+                // Store data for use with the next line.
+                prev_line_indent = (is_tabs, count);
+            }
+
+            // Give more weight to tabs, because their presence is a very
+            // strong indicator.
+            histogram[0] *= 2;
+
+            histogram
+        };
+
+        // Find the most frequent indent, its frequency, and the frequency of
+        // the next-most frequent indent.
+        let indent = histogram
+            .iter()
+            .enumerate()
+            .max_by_key(|kv| kv.1)
+            .unwrap()
+            .0;
+        let indent_freq = histogram[indent];
+        let indent_freq_2 = *histogram
+            .iter()
+            .enumerate()
+            .filter(|kv| kv.0 != indent)
+            .map(|kv| kv.1)
+            .max()
+            .unwrap();
+
+        // Use the auto-detected result if we're confident enough in its
+        // accuracy, based on some heuristics.  Otherwise fall back to
+        // the language-based setting.
+        if indent_freq >= 1 && (indent_freq_2 as f64 / indent_freq as f64) < 0.66 {
```

Comment:
Have you seen Xi's algorithm? They look for the greatest common denominator and use that as the indent unit: https://github.com/xi-editor/xi-editor/blob/52bf34368e52369106d8eca6096a4839f7629e1e/rust/core-lib/src/whitespace.rs#L89-L109

Reply:
I hadn't seen their implementation specifically, but skimming through it now it looks to be roughly the same as an earlier approach I tried back in the day.  Although they're being a little more clever by only computing the GCD on the most common line indentations, the general issue with using GCD is that if you have a file like this:

```rust
fn foo() {
    call_something_else() // Indented 4 spaces.
      .map(|v| v.bar)     // Only indented 2 more spaces because
      .unwrap();          // it's not a code block.

    // More code that's indented normally.
    ...
    ...
    ...
    ...
}
```

It ends up thinking that the indentation style is 2 spaces, even though the real indentation style is 4 spaces.  Granted, code formatted by e.g. rustfmt won't have this problem.  But outside of the Rust world (and probably even occasionally within it) there are plenty of imperfect code files.  And ideally you want an algorithm that's pretty robust against that.

The advantage of the algorithm I've implemented is that it looks at indentation in a very similar way as people do (or at least I do...?), where it's looking for the moments when a block of code increases in indentation, and it uses those increases to figure things out.

(It could also look at decreases in indentation, and I did that in a previous iteration of this algorithm.  But that didn't work out so well for Python files... so that's why it only counts increases.)

I certainly won't claim this is the best possible algorithm, but unlike GCD it's pretty hard to confuse unless you actually just throw it a file with messed up mixed indentation.